Add validate address options (#804)
substrate-developer-hub/substrate-developer-hub.github.io: docs/knowledgebase/advanced/ss58-address-format.md

title: SS58 Address Format
---

SS58 is a simple address format designed for Substrate based chains, primarily 
inspired by the [Bitcoin Base58Check](https://en.bitcoin.it/wiki/Base58Check_encoding) format with a few alterations.
Substrate developers are free to use the address format of their choice, but this serves as a robust default. 

> The purpose of this article is to outline existing solutions to validating a SS58 address.

Reviewer: It's not the purpose of this article. It's the purpose of this PR.

Reviewer 2: My bad, this was I mistake. I'll just suggest my changes in review mode.

  Account ID:         0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  SS58 Address:       12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU

# An invalid address.
$ subkey inspect "12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZUInvalidAddress"
Invalid phrase/URI given
```

### Polkadot.js

For JavaScript projects, the best solution is to use the functions for verifying an address 
provided by the [Polkadot.js API](https://polkadot.js.org/docs/api/). 
Here's what that would look like:

```javascript
// Import Polkadot.js API dependencies.
const { decodeAddress, encodeAddress } = require("@polkadot/keyring");
const { hexToU8a, isHex } = require("@polkadot/util");

// Specify an address to test.
const address = "<addressToTest>";

// Check address.
const isValidSubstrateAddress = () => {
  try {
    encodeAddress(isHex(address) ? hexToU8a(address) : decodeAddress(address));
    return true;
  } catch (error) {
    return false;
  }
};

// Query result.
const isValid = isValidSubstrateAddress();
console.log(isValid);
```

### Community Built

As for other solutions, here is a list of community maintained projects for developers 

> looking for address verification tools in different languages: 

Reviewer: Revert this. These are not "address verification tools".

- [Python Substrate Interface](https://polkascan.github.io/py-substrate-interface/#substrateinterface.Keypair) - developed by Polkascan.

- [Go Substrate RPC client](https://github.com/centrifuge/go-substrate-rpc-client) - developed by Centrifuge.

- [C++ Polkadot Substrate API](https://github.com/usetech-llc/polkadot_api_cpp) - developed by Usetech.

- [.Net API](https://github.com/usetech-llc/polkadot_api_dotnet) - developed by Usetech.

- [Go implemented utilities](https://github.com/itering/subscan-essentials/tree/master/util) - developed by Subscan.

## Learn more

- [SS58 Transform tool](https://polkadot.subscan.io/tools/ss58_transform)

> - [Polkadot-js API on GitHub](https://github.com/polkadot-js/api)

Reviewer: Why?

chain. Different chains have different means of identifying accounts. SS58 is designed to be
extensible for this reason.

The living specification for the SS58 address format can be found on the Substrate GitHub wiki:

https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)

## Validating Addresses

There are several ways to verify that a value is a valid SS58 address.

> ### Subkey

Reviewer: This section should show what happens when you input an _invalid_ address.

subcommand, which accepts the seed phrase, a hex-encoded address, or an SS58 address as the input
URI. If the input is a valid address, it will return a list containing the corresponding public
key (hex), account ID, and SS58 values.

Subkey currently works for keys generated using: sr25519, ed25519, and secp256k1.

> **NOTE:** If you input a valid SS58 value, Subkey will also return a network ID/version value
> that indicates which network the address has been encoded for.

```bash
$ subkey inspect "12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU"
Public Key URI `12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU` is account:
  Network ID/version: polkadot
  Public key (hex):   0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  Account ID:         0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  SS58 Address:       12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU
```

### Polkadot.js

For verifying an address in your JavaScript projects, you can utilize the functions built
into the [Polkadot.js API](https://github.com/polkadot-js/api/), which supports ed25519 and

> sr25519 key types.

Reviewer: 1. I'm not sure they key types are relevant here.
2. Are you sure it does not support secp256k1?

Author: Yeah, I'm not the greatest on the cryptography algos, but yeah on second glance I believe it does. Removing the key types in any case.

```javascript
const { decodeAddress, encodeAddress } = require("@polkadot/keyring");
const { hexToU8a, isHex } = require("@polkadot/util");

const address = "<addressToTest>";

const isValidAddressPolkadotAddress = () => {
  try {
    encodeAddress(isHex(address) ? hexToU8a(address) : decodeAddress(address));

    return true;
  } catch (error) {
    return false;
  }
};

const isValid = isValidAddressPolkadotAddress();

console.log(isValid);
```

> ### Subscan

Reviewer: Should go down into the community section.

You can also verify an address is valid by using the web tool on
[Subscan](https://polkadot.subscan.io/tools/ss58_transform). They have also made the
[essentials](https://github.com/itering/subscan-essentials/tree/78de8d163a3543a217f0cb3d48c6b9816bb5a231)
behind their web interface open source so you can take advantage of their
[Go implemented utilities](https://github.com/itering/subscan-essentials/tree/78de8d163a3543a217f0cb3d48c6b9816bb5a231/util)

> to suit your needs

Reviewer: ```suggestion
to suit your needs.
```

> ### Custom Solution

Reviewer: This whole section has a lot of conditional fluff. "You may", "You could", etc. I would just add a little intro that there are community-developed libraries in other languages and then add a bullet point list of them.

Author: Ok I am trimming it way down

If you wish to make your own custom solution for validating addresses you will need to consider
how an address is created. You may find some helpful documentation in the
[Substrate Github Wiki](<https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)>)

> pages, as well as in this [article](https://hackmd.io/@gavwood/r1jTRX2Zr#3-Working-with-SS58-and-account-addresses)

Reviewer: This is quite old and I don't like linking to hackmd anyway. Info that is useful/correct should go into the appropriate devhub page.

Author: removed these references for a more general community section description

chain. Different chains have different means of identifying accounts. SS58 is designed to be
extensible for this reason.

The living specification for the SS58 address format can be found on the Substrate GitHub wiki:

https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)

## Validating Addresses

There are several ways to verify that a value is a valid SS58 address.

### Subkey

You can use the [Subkey](https://substrate.dev/docs/en/knowledgebase/integrate/subkey) `inspect`
subcommand, which accepts the seed phrase, a hex-encoded private key, or an SS58 address as the input
URI. If the input is a valid address, it will return a list containing the corresponding public
key (hex), account ID, and SS58 values.

Subkey assumes that an address is based on a public/private keypair. In the case of inspecting an
address, it will return the 32 byte account ID. Not all addresses in Substrate-based networks are
based on keys.

> **NOTE:** If you input a valid SS58 value, Subkey will also return a network ID/version value

> > that indicates which network the address has been encoded for.

Reviewer: ```suggestion
> that indicates for which network the address has been encoded.
```

```bash
$ subkey inspect "12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU"
Public Key URI `12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU` is account:
  Network ID/version: polkadot
  Public key (hex):   0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  Account ID:         0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  SS58 Address:       12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU

> ```

Author: Not sure what you mean. It's a one-line command with the rest being an example of what the output would look like.

> ```

Reviewer 2: This was already resolved by @joepetrowski right above -- it was just about combining the bash commands + examples :)

> If you input an invlaid address you will get:

Reviewer: ```suggestion
If you input an invalid address you will get:
```

> ```
> 
> If you input an invalid address you will get:
> 
> ```bash
> $ subkey inspect "12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZUInvalidAddress"
> Invalid phrase/URI given
> ```
> 

Reviewer 2: ```suggestion

# An invalid address.
$ subkey inspect "12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZUInvalidAddress"
Invalid phrase/URI given
```
```

### Polkadot.js

For verifying an address in your JavaScript projects, you can utilize the functions built
into the [Polkadot.js API](https://github.com/polkadot-js/api/).

```javascript
const { decodeAddress, encodeAddress } = require("@polkadot/keyring");
const { hexToU8a, isHex } = require("@polkadot/util");

const address = "<addressToTest>";

> const isValidAddressPolkadotAddress = () => {

Reviewer: I think `isValidSubstrateAddress`

  try {
    encodeAddress(isHex(address) ? hexToU8a(address) : decodeAddress(address));

    return true;
  } catch (error) {
    return false;
  }
};

const isValid = isValidAddressPolkadotAddress();

console.log(isValid);
```

> ### Solutions from the Community

Reviewer 2: If we could knock off a few words here that would be nice. Something like "Community Built" or "Community Solutions".

If you are looking to build custom solutions for validating addresses in the language of your choice,

> you will find a list of projects by our community below to use as a starting point:

Reviewer: We can make it more concise, but it still needs to be accurate. This list is just of other Substrate API tools and are not specific to validating addresses.

> you will find a list of projects by our community below to use as a starting point:

Author: Ok made some changes to the wording here.

- [Python Substrate Interface](https://github.com/polkascan/py-substrate-interface) - developed by Polkascan.
  [Documentation](https://polkascan.github.io/py-substrate-interface/#substrateinterface.Keypair).

> - [Go Substrate RPC client](https://github.com/centrifuge/go-substrate-rpc-client/tree/3e974433f8417e386b033fb64a6ac4971f02c737) - developed by Centrifuge.

Reviewer: ```suggestion
- [Go Substrate RPC client](https://github.com/centrifuge/go-substrate-rpc-client) - developed by Centrifuge.
```

- [C++ Polkadot Substrate API](https://github.com/usetech-llc/polkadot_api_cpp) - developed by Usetech.

- [.Net API](https://github.com/usetech-llc/polkadot_api_dotnet) - developed by Usetech.

> - [Go implemented utilities](https://github.com/itering/subscan-essentials) - developed by Subscan.

Reviewer 2: For example: 

```suggestion
- [Go implemented utilities](https://github.com/itering/subscan-essentials/tree/master/util) - developed by Subscan.
```

Reviewer: I generally agree, but I like to point to somewhere with docs. This link just drops you into some code files, so I like to err on the side of going up the tree until there are some docs that will provide context about the library and how to use it.

> - [Go implemented utilities](https://github.com/itering/subscan-essentials/tree/78de8d163a3543a217f0cb3d48c6b9816bb5a231/util) - developed by Subscan.

Reviewer: I would just point to the page with the docs on it.
```suggestion
- [Go implemented utilities](https://github.com/itering/subscan-essentials) - developed by Subscan.
```

> - [SS58 Transform](https://polkadot.subscan.io/tools/ss58_transform) - to manually verify an address and convert between registered prefix values.

Reviewer: A "Learn More" section is fine, but Learn More sections typically point to other Parity hosted resources (like other KB articles or reference docs for Substrate/Polkadot JS). Subscan's transform tool should obviously be in the community built section.

> - [SS58 Transform](https://polkadot.subscan.io/tools/ss58_transform) - to manually verify an address.

Reviewer: ```suggestion
- [SS58 Transform](https://polkadot.subscan.io/tools/ss58_transform) - to manually verify an address and convert between registered prefix values.
```

chain. Different chains have different means of identifying accounts. SS58 is designed to be
extensible for this reason.

> The living specification for the SS58 address format can be found on the [Substrate GitHub wiki](<https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)>).

Reviewer: ```suggestion
The living specification for the SS58 address format can be found on the
[Substrate GitHub wiki](<https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)>).
```

The living specification for the SS-58 address format can be found on the Substrate GitHub wiki:

https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)

## Validating addresses

> There are several ways to verify that a value is a valid SS58 address before accepting it as input.

Reviewer 3: ```suggestion
There are several ways to verify that a value is a valid SS58 address.
```

> The living specification for the SS-58 address format can be found on the Substrate GitHub wiki:

Reviewer 3: ```suggestion
The living specification for the SS58 address format can be found on the Substrate GitHub wiki:
```

https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)

> ## Validating addresses

Reviewer 3: ```suggestion
## Validating Addresses
```

There are several ways to verify that a value is a valid SS58 address.

### Subkey

> You can use the [Subkey](https://substrate.dev/docs/en/knowledgebase/integrate/subkey) inspect

Reviewer 3: ```suggestion
You can use the [Subkey](https://substrate.dev/docs/en/knowledgebase/integrate/subkey) `inspect`
```

> subcommand which will accept the seed phrase, an address as hex or an SS58 address as the input

Reviewer 3: ```suggestion
subcommand, which accepts the seed phrase, a hex-encoded address, or an SS58 address as the input
```

> URI. If the input is a valid address, it will return a list containing the corresponding Public
> key (hex), Account ID, and SS58 values.

Reviewer 3: ```suggestion
URI. If the input is a valid address, it will return a list containing the corresponding public
key (hex), account ID, and SS58 values.
```

> Subkey currently works for keys generated using: sr25519, ed25519, and secp256k1.

Reviewer 3: Can we link out to reference materials for these?

> > **NOTE:** If you input a valid SS58 value Subkey will also return a Network ID/version value which indicates which network the address has been encoded for.

Reviewer 3: Line length, please.

```suggestion
> **NOTE:** If you input a valid SS58 value, Subkey will also return a network ID/version value that indicates which network the address has been encoded for.
```

```bash
$ subkey inspect "12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU"
Public Key URI `12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU` is account:
  Network ID/version: polkadot
  Public key (hex):   0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  Account ID:         0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  SS58 Address:       12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU
```

### Polkadot.js

For verifying an address in your JavaScript projects, you can utilize the functions built
into the [Polkadot.js API](https://github.com/polkadot-js/api/), which supports ed25519 and
sr25519 key types.

```javascript
const { decodeAddress, encodeAddress } = require("@polkadot/keyring");
const { hexToU8a, isHex } = require("@polkadot/util");

const address = "<addressToTest>";

const isValidAddressPolkadotAddress = () => {
  try {
    encodeAddress(isHex(address) ? hexToU8a(address) : decodeAddress(address));

    return true;
  } catch (error) {
    return false;
  }
};

const isValid = isValidAddressPolkadotAddress();

console.log(isValid);

> ```

Reviewer 3: Nice!

The living specification for the SS-58 address format can be found on the Substrate GitHub wiki:

https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58)

## Validating addresses

There are several ways to verify that a value is a valid SS58 address.

### Subkey

You can use the [Subkey](https://substrate.dev/docs/en/knowledgebase/integrate/subkey) inspect
subcommand which will accept the seed phrase, an address as hex or an SS58 address as the input
URI. If the input is a valid address, it will return a list containing the corresponding Public
key (hex), Account ID, and SS58 values.

Subkey currently works for keys generated using: sr25519, ed25519, and secp256k1.

> > **NOTE:** If you input a valid SS58 value Subkey will also return a Network ID/version value

Reviewer 3: ```suggestion
> **NOTE:** If you input a valid SS58 value, Subkey will also return a network ID/version value
```

> > which indicates which network the address has been encoded for.

Reviewer 3: ```suggestion
> that indicates which network the address has been encoded for.
```

```bash
$ subkey inspect "12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU"
Public Key URI `12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU` is account:
  Network ID/version: polkadot
  Public key (hex):   0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  Account ID:         0x46ebddef8cd9bb167dc30878d7113b7e168e6f0646beffd77d69d39bad76b47a
  SS58 Address:       12bzRJfh7arnnfPPUZHeJUaE62QLEwhK48QnH9LXeK2m1iZU
```

### Polkadot.js

For verifying an address in your JavaScript projects, you can utilize the functions built
into the [Polkadot.js API](https://github.com/polkadot-js/api/), which supports ed25519 and
sr25519 key types.

```javascript
const { decodeAddress, encodeAddress } = require("@polkadot/keyring");
const { hexToU8a, isHex } = require("@polkadot/util");

const address = "<addressToTest>";

const isValidAddressPolkadotAddress = () => {
  try {
    encodeAddress(isHex(address) ? hexToU8a(address) : decodeAddress(address));

    return true;
  } catch (error) {
    return false;
  }
};

const isValid = isValidAddressPolkadotAddress();

console.log(isValid);
```

### Subscan

> You can also manually convert and verify an address is valid by using the web tool on

Reviewer 3: ```suggestion
You can also verify an address is valid by using the web tool on
```

[Subscan](https://polkadot.subscan.io/tools/ss58_transform). They have also made the
[essentials](https://github.com/itering/subscan-essentials/tree/78de8d163a3543a217f0cb3d48c6b9816bb5a231)
behind their web interface open source so you can take advantage of their

> [Go implemented Utilities](https://github.com/itering/subscan-essentials/tree/78de8d163a3543a217f0cb3d48c6b9816bb5a231/util)

Reviewer 3: ```suggestion
[Go implemented utilities](https://github.com/itering/subscan-essentials/tree/78de8d163a3543a217f0cb3d48c6b9816bb5a231/util)
```

by Gavin Wood.

##### Python

For some inspiration on developing a solution in Python, you may find the
[Python Substrate Interface](https://github.com/polkascan/py-substrate-interface),developed by
Polkascan helpful. Here is where you can find their
[documentation](https://polkascan.github.io/py-substrate-interface/#substrateinterface.Keypair).

##### Go

Centrifuge has a client [implemented in Go](https://github.com/centrifuge/go-substrate-rpc-client/tree/3e974433f8417e386b033fb64a6ac4971f02c737)
that has Subkey as a dependency that you could use as a
[starting point](https://github.com/centrifuge/go-substrate-rpc-client/tree/3e974433f8417e386b033fb64a6ac4971f02c737).

##### C++

To help you along the way with a C++ solution you might want to look at
[Usetech's API implementation](https://github.com/usetech-llc/polkadot_api_cpp). Here is where they
[encode and decode addresses](https://github.com/usetech-llc/polkadot_api_cpp/blob/master/src/utils/address.cpp).

##### C#/.Net

Usetech also has a [.Net implementation](https://github.com/usetech-llc/polkadot_api_dotnet)
which may serve as a reference for dealing with addresses in C#. Here is where you can find

> their [address utilities](https://github.com/usetech-llc/polkadot_api_dotnet/blob/master/Polkadot/src/Utils/Address.cs).

Reviewer 3: These sections do not feel super complete/useful to me. I would rather just say something like "please refer to the `[client libraries](link)` for examples of working with the SS58 address format in a number of programming languages" or something.

Author: Yeah, I get what you mean. The reason why I included them was that it took me some digging in each of the client libraries just to find the areas where they deal with addresses, I thought it would save folks the initial repo digest by giving them a link to the top-level client library, and then a link into where the address specific code is.

Reviewer 3: Yep, I definitely get where you're coming from. When I'm faced with these types of questions, I always err on the side of completeness and clarity...if the topic isn't specific enough to Substrate development to allow me to dive into it completely and with total clarity, I think it's best to omit it. There is already so much important stuff we need to create and maintain, that I think it's very important to be pretty restrictive with what we choose to document.

Author: Trimmed down to repo links to provide some direction
